### PR TITLE
[tune] Fix HyperOptSearch dropping tune.choice categories that are constant dicts

### LINKED ISSUE #49507: missing search parameter in HyperOptSearch
### What happened + What you expected to happen

I have a set of pre-tuned hyperparameters. Now I want to tune other parameters while randomly sampling the pre-tuned parameters.  When using HyperOptSearch, those pre-tuned parameters are somehow lost during the new tuning process. If I remove "search_algo = HyperOptSearch(...)', the pre-tuned parameters do show up. 

### Versions / Dependencies

ray 2.40. 0

### Reproduction script

      import ray
      from ray import tune
      import numpy as np
      from ray.tune.search.hyperopt import HyperOptSearch

      # Define the existing hyperparameters as a list of dictionaries
      existing_hyperparameters = [
          {"learning_rate": 0.01, "batch_size": 32},
          {"learning_rate": 0.1, "batch_size": 2},
      ]

      # Define the new extra hyperparameter to optimize
      new_hyperparameter = {"new_param": tune.uniform(0.1, 10)}

      # Define the search space
      search_space = {
          "existing_hyperparameters": tune.choice(existing_hyperparameters),
          **new_hyperparameter,
      }

      # Define the objective function to optimize
      def objective(config):
          existing_hyperparameters = config["existing_hyperparameters"]
          new_param = config["new_param"]
          # Simulate some objective function evaluation
          return {"result": 0}

      # Initialize Ray
      ray.init()

      # Create a Tuner object
      tuner = tune.Tuner(
          objective,
          param_space=search_space,
          tune_config=tune.TuneConfig(
              num_samples=1,
              #search_alg=HyperOptSearch( metric="result", mode="max"),
              )
      )

      # Run the hyperparameter search
      results = tuner.fit()
      # Shutdown Ray

      ray.shutdown()

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by vineethsaivs**
## Description

`tune.choice()` over a list of constant dicts (a common way to express pre-tuned hyperparameter sets) silently loses every parameter with `HyperOptSearch`:

```python
config = {"space": tune.choice([{"a": 1, "b": 2}, {"a": 3, "b": 4}])}
searcher = HyperOptSearch(space=config, metric="a", mode="max")
searcher.suggest("0")   # {"space": {}}  <- all params dropped
```

Root cause: the `Categorical` branch of `HyperOptSearch.convert_search_space` recurses into each dict category, and a dict containing only plain constants has no `Domain`/grid vars, so the recursion hits the `if not domain_vars and not grid_vars: return {}` early return. `hp.choice` is then built over empty dicts. The same search space works with the default variant generator.

Hyperopt natively accepts constant dicts as choice categories (they are wrapped as literals by `as_apply`), so the fix falls back to the original category when the recursive conversion comes back empty: `convert_search_space(category, prefix=par) or category`. Dict categories that do contain `Domain` objects still convert exactly as before (the recursion returns non-empty), `grid_search` inside a category still raises, and the top-level all-constants `RuntimeError` path is untouched.

## Related issues

Fixes #49507

## Additional information

Added `testConvertHyperOptChoiceOfConstantDicts` to `SearchSpaceTest` in `python/ray/tune/tests/test_sample.py`, asserting a suggested `space` is one of the original dicts. It fails before this change (`suggest()` returns `{"space": {}}`) and passes after. The neighboring `testConvertHyperOptConstant` and `testConvertHyperOptNested` (full `tune.run`, 10 trials) still pass. `black` clean on both files; commit is DCO signed-off.

### CODE REVIEW

File: python/ray/tune/search/hyperopt/hyperopt_search.py [hunk 1]
```diff
@@ -522,9 +522,13 @@ def resolve_value(par: str, domain: Domain) -> Any:
...
                         [
                             (
+                                # A dict of plain constants converts to {}, but
+                                # hyperopt accepts constant dicts as choice
+                                # categories, so pass the original through.
                                 HyperOptSearch.convert_search_space(
                                     category, prefix=par
                                 )
+                                or category
                                 if isinstance(category, dict)
```
[the remark this answers is not in the document]
Comment (vineethsaivs) [reply python/ray/tune/search/hyperopt/hyperopt_search.py]: Agreed that the nested ternary is hard to read, but it predates this PR: the change here is the two-token `or category` on the existing `isinstance(category, dict)` branch, plus a comment. Rewriting the whole comprehension into a helper would turn a one-line behavior fix into a refactor of untouched code, which makes the fix harder to review and to backport.

Happy to do the `convert_category` extraction as a separate follow-up, or to fold it into this PR if a maintainer would rather see both together.

[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply python/ray/tune/search/hyperopt/hyperopt_search.py]: Update to this code, the nested loops are a pain to read
```
            elif isinstance(domain, Categorical):
                if isinstance(sampler, Uniform):
                    seq = []
                    for i, category in enumerate(domain.categories):
                        if isinstance(category, dict):
                            # Support constant dicts as choice categories
                            choice = HyperOptSearch.convert_search_space(
                                category, prefix=par
                            )
                            if choice:
                                seq.append(choice)
                            else:
                                seq.append(category)
                        elif (
                            isinstance(category, list)
                            and len(category) > 0
                            and isinstance(category[0], Domain)
                        ):
                            seq.append(
                                HyperOptSearch.convert_search_space(
                                    dict(enumerate(category)), prefix=f"{par}/{i}"
                                )
                            )
                        elif isinstance(category, Domain):
                            seq.append(resolve_value(f"{par}/{i}", category))
                        else:
                            seq.append(category)
                    return hpo.hp.choice(par, seq)
```

**Review by pseudo-rnd-thoughts — COMMENTED** [review]
Thanks for the PR, doesn't this solution only work if all the key-values within the dictionary are constant or variable. If you have a mixture of variables and constant then you still lose the constants. (I'm not sure if this is a valid config).

Also, lets partially implement gemini's comment

**Comment by vineethsaivs** [comment]
Thanks for the review. Done, I unrolled the nested ternary into the explicit loop you posted.

On the mixed constant/variable dict: it actually doesn't lose the constants. `convert_search_space` returns the full (deep-copied) `spec` with only the `Domain` entries replaced by hyperopt expressions; the empty-dict path (`if not domain_vars and not grid_vars: return {}`) is only taken when the category is entirely constant. So for a mixed category like `{"const": 5, "var": tune.uniform(0, 1)}` the call returns `{"const": 5, "var": <hp.uniform>}` and the constant survives. The `or category` fallback (now the `if choice: ... else: ...` in the loop) is only needed for the all-constant case, where the early return gives `{}`.

I added `testConvertHyperOptChoiceOfMixedConstantAndVariableDicts` to cover that: it suggests from a choice of a single mixed dict and asserts `const` stays `5` while `var` lands in `[0, 1]`.

**Review by pseudo-rnd-thoughts — APPROVED** [review]
Looks good, thanks for the update
